[RFC] fabric: Introduce new fi_trywait call

shefty opened the pull request:
Add a new call that applications must use when waiting on the
native wait object associated with an EQ or CQ.  The expected
behavior is for apps to do something like this:

    while (1) {
        if (fi_trywait(cq, ...) == 0)
            poll(cq_fd, ...);
 
       do {
            ret = fi_cq_read(cq, ...);
        } while (ret > 0);
    }

The trywait call is responsible for performing any steps needed
to safely block on the underlying fd using standard OS calls.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

shefty commented:
FIxes #1645 - or will once the header files and providers are updated.

bturrubiates commented on man/fi_fabric.3.md:
```diff
@@ -122,6 +134,58 @@ fi_tostr() will return a pointer to an internal libfabric buffer that
+
+The fi_trywait() call is used in conjunction with native operating
+system calls to block on wait objects, such as file descriptors.  It
+must be called, with a successful return value of 0, prior to
```
Why not say it will return `FI_SUCCESS`?

shefty replied:
Uhm, yeah, I should say that.

shefty commented:
This is compile tested only.  Changes from v1:

Updated the fi_trywait description, and moved it from the fi_fabric man page to fi_poll.  Added dummy man page link.

Disabled all providers from supporting FI_GETWAIT.  @bturrubiates @sungeunchoi @hppritcha @a-ilango @jithinjosepkl @goodell   Providers must define fi_trywait before re-enabling (with this approach).

Added fi_trywait definition to the header files and common framework.

Provided a quick implementation of fi_trywait for the UDP provider to illustrate how it could be implemented.  This needs to be tested, which will require a coordinated update to fabtests.

Fabtests fails with these changes, since the FI_GETWAIT fi_control operation returns -FI_ENOSYS.

shefty commented:
Missed tagging @j-xiong

shefty commented:
I think we're close to a final solution here.  We just need to get the function signature right.  :)  These are the options I can think of:

1. fi_trywait(fabric, fid_array, count) -- what this PR defines
2. fi_trywait(fid) -- single fid, looping must be done by app
3. Static inline of option 1 implemented over option 2.
4. fi_trywait(fabric, fid_set, count)
    fid_set would be similar to struct pollfd.

Hopefully those options make sense.

goodell commented:
> These are the options I can think of:
> 
> 1. fi_trywait(fabric, fid_array, count) -- what this PR defines

I'm still grokking and reviewing this PR in general, but my instinct is to go with (1) for now.

goodell commented on man/fi_poll.3.md:
```diff
@@ -195,6 +204,62 @@ being closed, otherwise the call will return -FI_EBUSY.
+blocking on a native wait object, or the wait object may not be
+signaled, resulting in an application hang.  The following
+psuedo-code demonstrates the use of fi_trywait in conjunction with
+the OS select call.
```
nit: "OS select call" --> "OS select(2) call"

shefty replied:
> nit: "OS select call" --> "OS select(2) call"

Will be in next update

goodell commented on man/fi_poll.3.md:
```diff
@@ -195,6 +204,62 @@ being closed, otherwise the call will return -FI_EBUSY.
+that will be passed to the native wait routine, the corresponding
+fabric descriptor should first be passed to fi_trywait.  All fabric
+descriptors passed into a single fi_trywait call must make use of the
+same underlying wait object.
```
"All fabric descriptors passed into a single fi_trywait call must make use of the same underlying wait object."

I think this is a strange restriction, since for non-wait-set wait objects, there's no way for an application to request that, e.g., two CQs use the same wait object under the hood, right?  And for wait sets we would be force to pass at most a single wait set by the same restriction.  So I think the above restriction implies that the `count` argument to `fi_trywait` will always be `1`.

goodell replied:
On the call today @shefty indicated that the intended meaning here is that rather than all fids having the same underlying wait object, instead they should have the same underlying wait object **type**.  E.g., all mutexes or all file descriptors.  This is a reasonable restriction to me, and one that we could always relax later if we found a good reason to do so.

bturrubiates replied:
> This is a reasonable restriction to me, and one that we could always relax later if we found a good reason to do so.

What's the reasoning behind this restriction? Is it from a simplicity, usability, or performance perspective?

shefty replied:
> What's the reasoning behind this restriction? Is it from a simplicity,
> usability, or performance perspective?

There's no application need driving more functionality.  So this potentially makes things simpler for the provider.

shefty commented:
Closing PR - I will open a new request with the requested updates, plus support for fi_trywait from the UDP and sockets provider.  This will be paired with a fabtests update.
